### PR TITLE
Supports Task Cancellations

### PR DESCRIPTION
**Pull request by sake402**
Add Task Cancellation supports
## Types of changes

- [x] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/master/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: nanoFramework.System.Threading/System/Runtime/CompilerServices/AsyncTaskMethodBuilder.cs [hunk 1]
```diff
@@ -9,6 +9,108 @@
...
+        {
+            Debug.WriteLine($"AsyncTaskMethodBuilder:SetException");
+            task?.CompleteWithException(exception);
+        }
+        /// <summary>
+        /// Marks the task as successfully completed.
+        /// </summary>
+        public void SetResult()
+        {
+            Debug.WriteLine($"AsyncTaskMethodBuilder:SetResult");
```
Comment (josesimoes) [inline nanoFramework.System.Threading/System/Runtime/CompilerServices/AsyncTaskMethodBuilder.cs:90]: Apart from the debug output nothing else is happening, is this the intend behaviour?

File: nanoFramework.System.Threading/System/Runtime/CompilerServices/AsyncTaskMethodBuilder.cs [hunk 2]
```diff
@@ -9,6 +9,108 @@
...
+        }
+        /// <summary>
+        /// Associates the builder with the specified state machine.
+        /// </summary>
+        /// <param name="stateMachine"></param>
+        public void SetStateMachine(IAsyncStateMachine stateMachine)
+        {
+            Debug.WriteLine($"AsyncTaskMethodBuilder:SetStateMachine");
```
Comment (josesimoes) [inline nanoFramework.System.Threading/System/Runtime/CompilerServices/AsyncTaskMethodBuilder.cs:98]: Same here

File: nanoFramework.System.Threading/Threading/Internal/ThreadWorker.cs [hunk 3]
```diff
@@ -38,28 +39,29 @@ void Start()
...
-                        {
-                            cantSuspend = true;
-                            while (callback == null)
-                            {
-                                Thread.Sleep(100);
-                            }
-                        }
-                    }
+                    run.WaitOne();
+                    //if (cantSuspend)//if platform doesnt support suspend, use polling
```
Comment (josesimoes) [inline nanoFramework.System.Threading/Threading/Internal/ThreadWorker.cs:43]: If this code isn't needed anymore better remove it

File: nanoFramework.System.Threading/Threading/Internal/ThreadWorker.cs [hunk 4]
```diff
@@ -76,8 +78,9 @@ public void Post(WaitCallback callback, object state)
             }
             else
             {
-                if (!cantSuspend)
-                    thread.Resume();
+                run.Set();
+                //if (!cantSuspend)
```
Comment (josesimoes) [inline nanoFramework.System.Threading/Threading/Internal/ThreadWorker.cs:82]: Same here.

File: nanoFramework.System.Threading/Threading/Tasks/Task.cs [hunk 5]
```diff
@@ -260,9 +306,14 @@ protected internal void CompleteWithException(Exception e)
...
+        /// <summary>
+        /// Get the result of the async operation. Block the calling thread until task completes
+        /// </summary>
         public void GetResult()
         {
-            while (!IsCompleted) ;
+            //while (!IsCompleted) ;
+            resultWaitHandle.WaitOne();
+            resultWaitHandle.Set(); //use it as one shot
             if (Exception != null)
```
Comment (josesimoes) [inline nanoFramework.System.Threading/Threading/Tasks/Task.cs:317]: Please add open and closing brackets, despite being a single line. That's the code style the project is following.

File: nanoFramework.System.Threading/Threading/Tasks/Task.cs [hunk 6]
```diff
@@ -287,8 +338,10 @@ public TResult Result
         {
             get
             {
-                // TODO check loop
-                while (!IsCompleted);
+                resultWaitHandle.WaitOne();
+                resultWaitHandle.Set(); //use it as one shot
+                //// TODO check loop
```
Comment (josesimoes) [inline nanoFramework.System.Threading/Threading/Tasks/Task.cs:343]: Remove commented code if not needed.

File: nanoFramework.System.Threading/Threading/ThreadPool.cs [hunk 7]
```diff
@@ -75,25 +79,26 @@ static ThreadWorker GetWorkerById(int id)
...
         {
-            ThreadWorker pool = null;
+            ThreadWorker worker = null;
             for (int i = 0; i < workers.Count; i++)
             {
-                pool = workers[i];
-                if (pool.IsFree)
-                    return pool;
+                worker = workers[i];
+                if (worker.IsFree)
```
Comment (josesimoes) [inline nanoFramework.System.Threading/Threading/ThreadPool.cs:86]: Open and close brackets.

File: nanoFramework.System.Threading/Threading/ThreadPool.cs [hunk 8]
```diff
@@ -193,7 +198,7 @@ public static bool QueueUserWorkItem(WaitCallback callBack, object state)
                 pool.Post(callBack, state);
                 return true;
             }
-            return pendingWorks.Put(new WorkItem(callBack, state));
+            return pendingWorks.Enqueue(new WorkItem(callBack, state));//queue a work item that is not bound to a specific thread context
```
Comment (josesimoes) [inline nanoFramework.System.Threading/Threading/ThreadPool.cs:201]: Comments inline with code, not after the code line.

File: nanoFramework.System.Threading/packages.config [hunk 9]
```diff
@@ -1,8 +1,5 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
-  <package id="Microsoft.Build.Tasks.Git" version="1.0.0" targetFramework="netnanoframework10" developmentDependency="true" />
```
Comment (josesimoes) [inline nanoFramework.System.Threading/packages.config:3]: These NuGet packages are meant to be on the packages file and are used in the build.
Please don't remove them.
Microsoft.Build.Tasks.Git
Microsoft.SourceLink.Common
Microsoft.SourceLink.GitHub
Nerdbank.GitVersioning

**Review by josesimoes — CHANGES_REQUESTED** [review]

File: nanoFramework.System.Threading/System/Runtime/CompilerServices/AsyncTaskMethodBuilder.cs (hunk 2 above)
Comment (sake402) [reply nanoFramework.System.Threading/System/Runtime/CompilerServices/AsyncTaskMethodBuilder.cs:98]: This is the intended behavior for now as we do not use the stateMachine outside of this function.

**Comment by josesimoes** [comment]
@sake402 I've fixed the references on the project and it's now building.

Some XML comments are still missing but that's OK for now.

I'm merging this now to allow further progress on this.

**Comment by josesimoes** [comment]
@sake402 also: please look at the failures from SonarCloud. Either fix them or wrap them in ignore clauses if they are meant to be that way.

**Pull request merged**